### PR TITLE
Run standalone dataloader and checkpointer on CPUs.

### PR DESCRIPTION
**Pull request by RoshaniN**
- Introducing JAX distributed initialize for CPU multihost environment. (Thanks to @priyanka-ganesha for finding the required CPU settings.)
- Added retry logic to ensure processes connect to the coordinator once it is up.
- Updating the workflow test names.
- Fixing bug in standalone_dataloader where Mesh was being initialized twice. Bug was introduced by commit http://shortn/_Q5ITxrra63 .
- Introducing 'hint_hardware' in the config.

### CODE REVIEW

File: MaxText/standalone_dataloader.py
Comment (RoshaniN) [inline MaxText/standalone_dataloader.py:66]: @rwitten @gobbleturk I think this line needs to be specific to CPUs (v4 runner tests are failing due to this line)  as these JAX variables will only be set for CPUs.

Is there a check that you could recommend?

Comment (rwitten) [reply MaxText/standalone_dataloader.py:66]: This is a Matt problem!

Comment (RoshaniN) [reply MaxText/standalone_dataloader.py:66]: Sure, will wait for his response, thanks Rafi!

Comment (RoshaniN) [reply MaxText/standalone_dataloader.py]: Thanks for the offline advise, Matt! @gobbleturk 

I moved the jax distributed initialize for CPUs to max utils. I also tested that async checkpointer now works for CPUs, with this changes.

Logs: 
with async checkpointer - https://cloudlogging.app.goo.gl/V7Qog8rJvRDgJRci8 
with sync checkpointer - https://cloudlogging.app.goo.gl/XzxRKqBYD2tXxW8g8 


Test failures are due to "Install dependencies" step, probably unrelated to my changes.

File: MaxText/max_utils.py
Comment (gobbleturk) [inline MaxText/max_utils.py:198]: nit: 

return os.environ.get("JAX_COORDINATOR_ADDRESS") is not None

instead of if statement

File: MaxText/standalone_dataloader.py
Comment (RoshaniN) [reply MaxText/standalone_dataloader.py]: https://github.com/google/maxtext/pull/399 will unblock the tests, I will rebase onto it shortly.

**Review by gobbleturk — APPROVED** [review]

File: MaxText/max_utils.py
Comment (RoshaniN) [reply MaxText/max_utils.py:198]: Makes sense, thanks!

File: MaxText/pyconfig.py
Comment (rwitten) [inline MaxText/pyconfig.py:172]: move this logic into funciton. you can early exit if your criteria rre met.

File: MaxText/standalone_dataloader.py
Comment (rwitten) [inline MaxText/standalone_dataloader.py:67]: If you're getting rid of flags, maybe get rid of this one.

**Review by rwitten — APPROVED** [review]
Little nits, please fix them before merging.

**Comment by RoshaniN** [comment]
I noticed that the dataloader is timing out and the connection to the JAX coordinator is shutdown. Hence moving this PR to draft again.

Dataloader was working fine previously with these changes, will check what changed.

Update - standalone_dataloader was updated in http://shortn/_Q5ITxrra63 , where Mesh was being initialized twice, hence the timeouts. Fixed now.

File: MaxText/pyconfig.py
Comment (RoshaniN) [reply MaxText/pyconfig.py]: Done, thanks

File: MaxText/standalone_dataloader.py
Comment (RoshaniN) [reply MaxText/standalone_dataloader.py]: Done, thanks

File: MaxText/max_utils.py
Comment (rwitten) [inline MaxText/max_utils.py:223]: sorry but let's just call this "hardware"

File: MaxText/standalone_checkpointer.py
Comment (rwitten) [inline MaxText/standalone_checkpointer.py:97]: you dont need this cache! just get rid of it totally!

File: MaxText/configs/base.yml
Comment (rwitten) [inline MaxText/configs/base.yml:99]: call it hardware

**Review by rwitten — APPROVED** [review]
Little nit but giving approval. Please deal with nit before merging

**Pull request merged**
